### PR TITLE
New heterogeneous atmosphere design

### PR DESCRIPTION
**Pull request by nollety**
Resolves eradiate/eradiate-issues#74.

# Description

This pull request introduces a new heterogeneous atmosphere design. 

Prerequisite pull requests:

* #111 
* #112

## Motivation

This new design was motivated by the need to integrate `ParticleLayer` objects in the heterogeneous atmosphere. The integration proved to be rather complex, for various reasons including:

* the level altitude meshes do not align
* the radiative properties must be "blended" together, including the phase functions

which is why this new design is suggested.

## New design layout

The new design is illustrated by the UML diagram below:

uml_diagram

* `MolecularAtmosphere` is the molecular part of the atmosphere, i.e. without particles. Each type of molecular atmospheres (e.g. AFGL1986) is associated to a specific class method (specified by `"construct"` in the corresponding kernel dict).
* `ParticleLayer` objects are 1D layers of particles (e.g. a cloud layer).
* `HeterogeneousAtmosphere` is an object that blends a `MolecularAtmosphere` object with one or more `ParticleLayer` objects. It has special methods that lets it align the altitude meshes of its components, compute the total collision coefficients and phase function blending ratios based on the individual radiative properties of each component, etc.

## Changes

* added a module `scenes.atmosphere._heterogeneous_new` where `HeterogeneousAtmosphereNew` is defined. We append __new_ or _New_ where necessary, in order to avoid breaking the existing implementation of `HeterogeneousAtmosphere`. Once the new design is completely implemented and tested, then it will be time to rename these. Corresponding tests are in `scenes/atmosphere/tests/test_heterogeneous_new.py`.
* added a module `_molecules` where `MolecularAtmosphere` is defined. Corresponding tests are in `scenes/atmosphere/tests/test_molecules.py`.
* fixed a typo to `_ROOTDIR` value in `eradiate/data/absorption_spectra.py`
* updated documentation

### `ParticleLayer`

`ParticleLayer` now inherits `Atmosphere` as indicated in the UML diagram above.
`ParticleLayer` is then now a scene element, which means

```python
from eradiate.contexts import KernelDictContext
from eradiate.scenes.atmosphere import ParticleLayer
from eradiate.units import unit_registry as ureg

ctx = KernelDictContext()
particle_layer = ParticleLayer(
    bottom = 0.0 * ureg.km,
    top = 1.5 * ureg.km,
    tau_550 = 0.1 * ureg.dimensionless,
    n_layers=9,
)
KernelDict.new(particle_layer, ctx=ctx)
```

produces a valid kernel dictionary.

### `MolecularAtmosphere`

```python
from eradiate.scenes.atmosphere import MolecularAtmosphere

molecular_atmosphere = MolecularAtmosphere()
```

### `HeterogeneousNewAtmosphere`

Empty atmosphere:
```python
from eradiate.scenes.atmosphere import HeterogeneousNewAtmosphere

atmosphere = HeterogeneousNewAtmosphere()
```

Atmosphere with only a molecular component:
```python
from eradiate.scenes.atmosphere import HeterogeneousNewAtmosphere, MolecularAtmosphere

atmosphere = HeterogeneousNewAtmosphere(molecular_atmosphere=MolecularAtmosphere())
```

Atmosphere with only a particles component:
```python
from eradiate.scenes.atmosphere import HeterogeneousNewAtmosphere, ParticleLayer

atmosphere = HeterogeneousNewAtmosphere(particle_layers=[ParticleLayer()])
```

Atmosphere with a molecular component and a particles component (one particle layer):
```python
from eradiate.scenes.atmosphere import HeterogeneousNewAtmosphere, MolecularAtmosphere, ParticleLayer
from eradiate.units import unit_registry as ureg

atmosphere = HeterogeneousNewAtmosphere(
    molecular_atmosphere=MolecularAtmosphere(id="molecules"),
    particle_layers=[ParticleLayer(id="particles")]
)
```

Atmosphere with a molecular component and a particles component (two particle layers):
```python
from eradiate.scenes.atmosphere import HeterogeneousNewAtmosphere, MolecularAtmosphere, ParticleLayer

atmosphere = HeterogeneousNewAtmosphere(
    molecular_atmosphere=MolecularAtmosphere(id="molecules"),
    particle_layers=[
        ParticleLayer(id="particles_1", bottom=0.0 * ureg.km, top=1.0 * ureg.km),
        ParticleLayer(id="particles_2", bottom=6.0 * ureg.km, top=7.0 * ureg.km),
    ]
)
```

#### :warning:  Disclaimer

There is no support for overlapping layers just yet.
Support for overlapping layers would be easier to add if the `blendphase` plugin would accept an arbitrary number of components.

# Checklist

- [x] The code follows the relevant coding guidelines
- [x] The code generates no new warnings
- [x] The code is appropriately documented
- [x] The code is tested to prove its function
- [x] The feature branch is rebased on the current state of the `main` branch
- [x] I give permission that the Eradiate project may redistribute my contributions under the terms of its license

### CODE REVIEW

**Comment by nollety** [comment]
@leroyvn, I'm getting a `RuntimeError: ​[PluginManager] Plugin "lut" not found!` when trying the use the lut plugin. Any idea what I'm doing wrong? I've run ninja to update the eradiate kernel.

Here is the full traceback:
```shell
_____________________________________________________________________________________________________________________________________________________ test_particle_load _____________________________________________________________________________________________________________________________________________________

mode_mono = None

    def test_particle_load(mode_mono) -> None:
        """Produces a kernel dictionary that can be loaded by the kernel."""
        ctx = KernelDictContext()
        particle_layer = ParticleLayer()
>       assert KernelDict.new(particle_layer, ctx=ctx).load() is not None

eradiate/scenes/atmosphere/tests/test_particles.py:16: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = {'type': 'scene', 'phase_atmosphere': {'type': 'lut', 'values': '277.2401,276.2332,273.3426,268.6447,262.4106,254.9236...0, 0, 500.5, 499.5],
 [0, 0, 0, 1]], 'bsdf': {'type': 'null'}, 'interior': {'type': 'ref', 'id': 'medium_atmosphere'}}}

    def load(self) -> "mitsuba.render.Scene":
        """
        Load kernel object from self.
    
        .. note:: Requires a valid selected operational mode.
    
        Returns → :class:`mitsuba.render.Scene`:
             Kernel object.
        """
        self.check()
        from mitsuba.core.xml import load_dict
    
>       return load_dict(self.data)
E       RuntimeError: ​[PluginManager] Plugin "lut" not found!

eradiate/scenes/core.py:140: RuntimeError
================================================================================================================================================== short test summary info ===================================================================================================================================================
FAILED eradiate/scenes/atmosphere/tests/test_particles.py::test_particle_load - RuntimeError: ​[PluginManager] Plugin "lut" not found!
```

**Comment by nollety** [comment]
Oh, okay, it's not `"lut"` (as mentioned in the docs) but `"lutphase"`. Thanks @schunkes!

File: eradiate/scenes/atmosphere/_heterogeneous_new.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+        if self.particle_layers is not None:
+            for particle_layer in self.particle_layers:
+                tops.append(particle_layer.top)
+
+        if len(tops) == 0:
+            return 10.0 * ureg.km  # dummy default
+        else:
+            return max(tops)
+
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_heterogeneous_new.py:108]: A lot of the branching used here would not be necessary if you would have a `components` property returning all components at once. Example:
```python
@property
def components(self):
    result = [self.molecular_atmosphere] if self.molecular_atmosphere else []
    result.extend(self.particle_layers)
    return result
```

File: eradiate/scenes/atmosphere/_heterogeneous_new.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            ids.append(instance.molecular_atmosphere.id)
+
+        if instance.particle_layers is not None:
+            for particle_layer in instance.particle_layers:
+                if not particle_layer.id in ids:
+                    ids.append(particle_layer.id)
+                else:
+                    raise ValueError(
+                        f"either two particle layers or one particle layer and the molecular atmosphere share the same id ('{ids[-1]}') which is forbidden."
+                    )
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_heterogeneous_new.py:68]: With `component` field (uses a set for optimal container lookup performance):
```python
ids = set()

for component in self.components:
    if not component.id in ids:
        ids.add(component.id)
    else:
        raise ValueError(
           f"while validating {attribute.name}: found duplicate component ID '{component.id}'; "
           f"all components (molecular atmosphere and particle layers) must have different IDs"
         )
```

File: eradiate/scenes/atmosphere/_heterogeneous_new.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+            bottoms.append(self.molecular_atmosphere.bottom)
+
+        if self.particle_layers is not None:
+            for particle_layer in self.particle_layers:
+                bottoms.append(particle_layer.bottom)
+
+        if len(bottoms) == 0:
+            return 0.0 * ureg.km  # dummy default
+        else:
+            return min(bottoms)
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_heterogeneous_new.py:91]: More concise with `component` property:
```python
bottoms = [component.bottom for component in self.components]
if bottoms:
    return min(bottoms)
else:
    return 0.0 * ureg.km
```

File: eradiate/scenes/atmosphere/_heterogeneous_new.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+            tops.append(self.molecular_atmosphere.top)
+
+        if self.particle_layers is not None:
+            for particle_layer in self.particle_layers:
+                tops.append(particle_layer.top)
+
+        if len(tops) == 0:
+            return 10.0 * ureg.km  # dummy default
+        else:
+            return max(tops)
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_heterogeneous_new.py:107]: With `component` property:
```python
tops = [component.top for component in self.components]
if bottoms:
    return max(tops)
else:
    return 10.0 * ureg.km
```

File: eradiate/scenes/atmosphere/_heterogeneous_new.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+            widths.append(self.molecular_atmosphere.eval_width(ctx=ctx))
+
+        if self.particle_layers is not None:
+            for particle_layer in self.particle_layers:
+                widths.append(particle_layer.eval_width(ctx=ctx))
+
+        if len(widths) == 0:
+            return 1000.0 * ureg.km  # dummy default
+        else:
+            return max(widths)
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_heterogeneous_new.py:122]: Ditto with `component` property.

File: eradiate/scenes/atmosphere/_molecules.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+
+    phase: PhaseFunction = documented(
+        attr.ib(
+            factory=lambda: RayleighPhaseFunction(),
+            converter=phase_function_factory.convert,
+            validator=attr.validators.instance_of(PhaseFunction),
+        )
+    )
+
+    has_absorption: Optional[bool] = documented(
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_molecules.py:68]: [`Optional[X]` is equivalent to `Union[X, None]`](https://docs.python.org/3/library/typing.html#typing.Optional); you don't implement this behaviour here, so your annotation should simply be:
```suggestion
    has_absorption: bool = documented(
```

File: eradiate/scenes/atmosphere/_molecules.py [hunk 7]
```diff
@@ -0,0 +1,440 @@
...
+            validator=attr.validators.instance_of(bool),
+        ),
+        doc="Absorption switch. If ``True``, the absorption coefficient is "
+        "computed. Else, the absorption coefficient is not computed and "
+        "instead set to zero.",
+        type="bool",
+        default="True",
+    )
+
+    has_scattering: Optional[bool] = documented(
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_molecules.py:81]:
```suggestion
    has_scattering: bool = documented(
```

File: eradiate/scenes/atmosphere/_molecules.py [hunk 8]
```diff
@@ -0,0 +1,440 @@
...
+    albedo_filename: str = documented(
+        attr.ib(
+            default="albedo.vol",
+            converter=str,
+            validator=attr.validators.instance_of(str),
+        ),
+        doc="Name of the albedo volume data file.",
+        type="str",
+        default='"albedo.vol"',
+    )
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_molecules.py:120]: Side remark: I'm still thinking about how to get rid of these, and I have a design solution :)

File: eradiate/scenes/atmosphere/tests/test_molecules.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    spectral_ctx = SpectralContext.new(wavelength=550.0)
+    ctx = KernelDictContext(spectral_ctx=spectral_ctx)
+    atmosphere = MolecularAtmosphere.afgl1986(
+        absorption_data_sets=afgl1986_test_absorption_data_sets
+    )
+    assert KernelDict.new(atmosphere, ctx=ctx).load() is not None
+
+
+@pytest.fixture
+def ussa76_approx_test_absorption_data_set() -> MutableMapping[str, str]:
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/tests/test_molecules.py:62]: Same comment as yesterday about type hints in test code: it's overkill, I'd rather have them removed (at least in fixtures and test functions; for utility functions, it's okay).

**Review by leroyvn — CHANGES_REQUESTED** [review]
Hi @nollety, nice PR :) My comments mostly have to do with style, nothing very critical. Once those are fixed, we can merge!

File: eradiate/data/absorption_spectra.py [hunk 10]
```diff
@@ -113,7 +113,7 @@ def _get_data_set_id(group, w_bin):
     return f"{absorber.value}-{engine.value}-{w_min}_{w_max}"
 
 
-_ROOT_DIR = "spectra/absorption"
+_ROOT_DIR = "resources/data/spectra/absorption"
```
[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/data/absorption_spectra.py:116]: I don't remember why I needed that; I just tried using a path relative to the data directory and all tests pass. I'll undo this change.

File: eradiate/scenes/atmosphere/_heterogeneous_new.py [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+        The given particle layer.
+
+    Returns → List[ParticleLayer]:
+        Overlapping particle layers.
+    """
+    overlapping = []
+    for layer in particle_layers:
+        if particle_layer.top >= layer.top > particle_layer.bottom:
+            overlapping.append(layer)
+        elif particle_layer.top > layer.bottom >= particle_layer.bottom:
```
[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/scenes/atmosphere/_heterogeneous_new.py:369]: No :)

File: eradiate/scenes/atmosphere/_homogeneous.py [hunk 12]
```diff
@@ -123,7 +123,7 @@ def top(self) -> pint.Quantity:
     def eval_width(self, ctx: Optional[KernelDictContext] = None) -> pint.Quantity:
         if self.width is AUTO:
             spectral_ctx = ctx.spectral_ctx if ctx is not None else None
-            return 10.0 / self.eval_sigma_s(spectral_ctx)
+            return 10.0 / self.eval_sigma_s(spectral_ctx).min()
```
[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/scenes/atmosphere/_homogeneous.py:126]: This `min()` call is irrelevant.

[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/scenes/atmosphere/_homogeneous.py:126]: I removed it

File: eradiate/scenes/atmosphere/_heterogeneous_new.py [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+            for particle_layer in instance.particle_layers:
+                if not particle_layer.id in ids:
+                    ids.append(particle_layer.id)
+                else:
+                    raise ValueError(
+                        f"either two particle layers or one particle layer and the molecular atmosphere share the same id ('{ids[-1]}') which is forbidden."
+                    )
+
+    # TODO: the user is probably interested in a shortcut to align the
+    # width of the particle layers with the width of the molecular atmosphere.
```
[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/scenes/atmosphere/_heterogeneous_new.py:71]: I'm not sure what you mean, I was thinking about something like that:
```python
align_components_widths: bool = documented(
        attr.ib(
            default=False,
            converter=bool,
            validator=attr.validators.instance_of(bool),
        ),
        doc="Set all components' widths to the width of the component with the largest width.",
        type="bool",
        default="False",
    )
```
What do you think?

[the remark this answers is not in the document]
Comment (nollety) [reply eradiate/scenes/atmosphere/_heterogeneous_new.py:82]: As discussed irl, all components widths must be aligned and we will only accept components with `width=AUTO`

File: eradiate/scenes/atmosphere/_molecules.py [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+
+    @property
+    def albedo_file(self) -> pathlib.Path:
+        return self.cache_dir / self.albedo_filename
+
+    @property
+    def sigma_t_file(self) -> pathlib.Path:
+        return self.cache_dir / self.sigma_t_filename
+
+    def kernel_phase(self, ctx: Optional[KernelDictContext]) -> MutableMapping:
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_molecules.py:235]: You can hint a parameter as `Optional` **iff** it has `None` as its default (other occurences below).

File: eradiate/scenes/atmosphere/_molecules.py [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+        radprops = self.radprops_profile.to_dataset(spectral_ctx=ctx.spectral_ctx)
+        albedo = to_quantity(radprops.albedo).m_as(uck.get("albedo"))
+        sigma_t = to_quantity(radprops.sigma_t).m_as(uck.get("collision_coefficient"))
+        write_binary_grid3d(
+            filename=str(self.albedo_file), values=albedo[np.newaxis, np.newaxis, ...]
+        )
+        write_binary_grid3d(
+            filename=str(self.sigma_t_file), values=sigma_t[np.newaxis, np.newaxis, ...]
+        )
+        if ctx.ref:
```
Comment (leroyvn) [inline eradiate/scenes/atmosphere/_molecules.py:261]: Typically here, `ctx` is required and cannot be `Optional`.

**Review by leroyvn — COMMENTED** [review]
Hi @nollety, one minor comment and we'll be good to go. When it's taken care of, we can merge.

**Pull request merged**